pkg/backend/httpstate/client/ai.go

```diff
@@ -0,0 +1,104 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"encoding/json"
+	"errors"
+	"strings"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+)
+
+// createSummarizeUpdateRequest creates a new CopilotSummarizeUpdateRequest with the given content and org ID
+func createSummarizeUpdateRequest(
+	lines []string,
+	orgID string,
+	model string,
+	maxSummaryLen int,
+) apitype.CopilotSummarizeUpdateRequest {
+	// Convert lines to a single string
+	content := strings.Join(lines, "\n")
+	content = TruncateWithMiddleOut(content, maxCopilotContentLength)
+
+	return apitype.CopilotSummarizeUpdateRequest{
+		State: apitype.CopilotState{
+			Client: apitype.CopilotClientState{
+				CloudContext: apitype.CopilotCloudContext{
+					OrgID: orgID,
+					URL:   "https://app.pulumi.com",
+				},
+			},
+		},
+		DirectSkillCall: apitype.CopilotDirectSkillCall{
+			Skill: "summarizeUpdate",
+			Params: apitype.CopilotSkillParams{
+				PulumiUpdateOutput: content,
+				Model:              model,
+				MaxLen:             maxSummaryLen,
+			},
+		},
+	}
+}
+
+// extractSummaryFromResponse parses the Copilot API response and extracts the summary content
+func extractSummaryFromResponse(copilotResp apitype.CopilotSummarizeUpdateResponse) (string, error) {
+	for _, msg := range copilotResp.ThreadMessages {
+		if msg.Role != "assistant" {
+			continue
+		}
+
+		// Handle the new format where content is a string directly
+		if msg.Kind == "response" {
+			// Unmarshal the RawMessage into a string
+			var contentStr string
+			if err := json.Unmarshal(msg.Content, &contentStr); err != nil {
+				// If it's not a simple string, it might be a raw JSON object
+				// Return it as a string representation
+				return string(msg.Content), nil
+			}
+			return contentStr, nil
+		}
+	}
+	return "", errors.New("no assistant message found in response")
+}
+
+// Maximum number of characters to send to Copilot.
+// We do this to avoid including a proper token counting library for now.
+// Tokens are 3-4 characters as a rough estimate. So this is 1000 tokens.
+const maxCopilotContentLength = 4000
+
+const truncationNotice = "... (truncated) ..."
+
+// TruncateWithMiddleOut takes a string and a maximum character count,
+// and returns a new string with content truncated from the middle if the total
+// character count exceeds maxChars. This preserves both the beginning and end
+// of the content while removing content from the middle.
+func TruncateWithMiddleOut(content string, maxChars int) string {
+	// If content is shorter than max or max is too small, return as is
+	if len(content) <= maxChars || maxChars <= len(truncationNotice) {
```

Comment:
:thinking: What about `TruncateWithMiddleOut(aStringWithAThousandCharacters, 6)`? Because 6 is less than the length of `truncationNotice`, we'll return a 1,000-character string? Wonder if we can do better.

Reply:
Good catch! Indeed. Will add some tests for maxChars < len(truncationNotes) 🤦 .

golang fuzzers? Probably..